### PR TITLE
Create API Authentication

### PR DESCRIPTION
**Pull request by OrangeSquirter**

### CODE REVIEW

File: clean.sh
Comment (vcaster) [inline clean.sh:1]: `docker compose down`  works any reason behind this

File: api-unit-tests.go
Comment (vcaster) [inline api-unit-tests.go:1]: We should move this to a `test/` dir

File: api/auth.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+func Auth(requiredRole string) gin.HandlerFunc{
+	return func(context *gin.Context) {
+		tokenString := context.GetHeader("Authorization")
+		if tokenString == "" {
+			context.JSON(401, gin.H{"error": "request does not contain an access token"})
+			context.Abort()
+			return
+		}
+		err:= ValidateToken(tokenString, requiredRole)
```
Comment (vcaster) [inline api/auth.go:15]: Where is this function

File: api/auth.go
Comment (vcaster) [inline api/auth.go:1]: Would be nice to have the `api/ `dir in `Web/api`

File: api/data.go
Comment (vcaster) [inline api/data.go:1]: Love this, but it seems like a duplicate with `data/data.go` you should be able to add/import functions from data.go

File: api/data.go [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+			&info.CallbackFrequency,
+			&info.CallbackJitter,
+		)
+	}
+	infoAppend = append(infoAppend, info)
+	logger.Logf(logger.Info, "%v\n", info)
+	return infoAppend
+}
+
+func createUserHandler(c *gin.Context) {
```
Comment (vcaster) [inline api/data.go:518]: Great! go functions needs to start with a Capital letter if you would like to use it in another file

File: api/data.go [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+
+    "github.com/gin-gonic/gin"
+	"github.com/PatronC2/Patron/helper"
+	"github.com/PatronC2/Patron/types"	
+	"github.com/PatronC2/Patron/lib/logger"	
+	"github.com/joho/godotenv"
+	_ "github.com/lib/pq"
+)
+
+type UserCreationRequest struct {
```
Comment (vcaster) [inline api/data.go:18]: you can make a `types/api.go` file for this

File: api/database.go
Comment (vcaster) [inline api/database.go:1]: Love the spliting this, we should test this

File: api/auth.go (hunk 1 above)
Comment (vcaster) [reply api/auth.go:15]: I see it in `api/jwt.go` we should add the import

File: api/jwt.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"errors"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+// pull that secret from env
+var jwtKey = []byte("patronsecret")
```
Comment (vcaster) [inline api/jwt.go:11]: Note: to add to install.sh

File: api/jwt.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+type JWTClaim struct {
```
Comment (vcaster) [inline api/jwt.go:13]: we could move this to `types/api.go`

File: api/jwt.go
Comment (vcaster) [inline api/jwt.go:1]: Beautiful

File: api/main.go
Comment (vcaster) [inline api/main.go:1]: lovely

File: api/models.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+    "fmt"
+
+    "golang.org/x/crypto/bcrypt"
+	"github.com/PatronC2/Patron/lib/logger"	
+)
+
+type User struct {
```
Comment (vcaster) [inline api/models.go:10]: we could move this to types/api.go

File: api/models.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    u.PasswordHash = string(hash)
+    return nil
+}
+
+func (u *User) CheckPassword(password string) error {
+    return bcrypt.CompareHashAndPassword([]byte(u.PasswordHash), []byte(password))
+}
+
+func getUserByUsername(username string) (*User, error) {
```
Comment (vcaster) [inline api/models.go:31]: Ohhhhh  noooooooooo
lol
we should move all sql functions to the data.go

File: data/data.go [hunk 8]
```diff
@@ -12,8 +12,16 @@ import (
 	"github.com/PatronC2/Patron/lib/logger"	
 	"github.com/joho/godotenv"
 	_ "github.com/lib/pq"
+	"golang.org/x/crypto/bcrypt"
 )
 
+type User struct {
```
Comment (vcaster) [inline data/data.go:18]: same for the types comment above

File: data/data.go [hunk 9]
```diff
@@ -616,3 +658,28 @@ func FetchOne(db *sql.DB, uuid string) []types.ConfigAgent {
 	logger.Logf(logger.Info, "%v\n", info)
 	return infoAppend
 }
+
+func (db *sql.DB, u *User) SetPassword(password string) error {
```
Comment (vcaster) [inline data/data.go:662]: interesting

File: data/main.go
Comment (vcaster) [inline data/main.go:1]: Should be on the `api/` dir

File: data/middleware.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
+package data
+
+import (
+    "github.com/gin-gonic/gin"
+    "github.com/dgrijalva/jwt-go"
+    "net/http"
+    "time"
+)
+
+var jwtSecret = []byte("your_secret_key")
```
Comment (vcaster) [inline data/middleware.go:10]: Note: to add to install.sh

File: docker-compose.yml
Comment (vcaster) [inline docker-compose.yml:1]: Lol, what happened

File: install.sh [hunk 11]
```diff
@@ -232,7 +232,7 @@ echo "REACT_APP_NGINX_PORT=$nginxport" >> .env
 echo "REACT_APP_NGINX_IP=$nginxip" >> .env
 echo "REACT_SERVER_IP=$reactclientip" >> .env
 echo "REACT_SERVER_PORT=$reactclientport" >> .env
-echo "BASIC_AUTH_PASS=$basicpass" >> .env
+echo "ADMIN_AUTH_PASS=$basicpass" >> .env
```
Comment (vcaster) [inline install.sh:235]: Let's still keep `BASIC_AUTH_PASS` and add `ADMIN_AUTH_PASS`

Comment (vcaster) [reply install.sh:235]: to fully deprecate `BASIC_AUTH` we need to remove the nginx docker definition and all reference to it

**Review by vcaster — COMMENTED** [review]
Yeah budeeeee, left some comments

File: api-unit-tests.go
Comment (OrangeSquirter) [reply api-unit-tests.go:1]: yep, coming in my next commit

File: api/auth.go (hunk 1 above)
Comment (OrangeSquirter) [reply api/auth.go:15]: moving stuff around still, probably all going to go into auth.go

File: api/data.go
Comment (OrangeSquirter) [reply api/data.go:1]: still working on this, one of the files will be removed

File: api/data.go (hunk 2 above)
Comment (OrangeSquirter) [reply api/data.go:518]: did not know this (I should probably read docs), but will adjust some stuff with this knowledge

File: data/main.go
Comment (OrangeSquirter) [reply data/main.go:1]: yep. need to just delete this file, it was a part of original failed testing

File: docker-compose.yml
Comment (OrangeSquirter) [reply docker-compose.yml:1]: commented out the containers for quick testing for now, still a work in progress

**Pull request merged**